integration/fuse/src/main/java/alluxio/fuse/file/FuseFileInStream.java

```diff
@@ -0,0 +1,137 @@
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.fuse.file;
+
+import alluxio.AlluxioURI;
+import alluxio.client.file.FileInStream;
+import alluxio.client.file.FileSystem;
+import alluxio.client.file.URIStatus;
+import alluxio.exception.AlluxioException;
+import alluxio.fuse.AlluxioFuseOpenUtils;
+import alluxio.fuse.AlluxioFuseUtils;
+
+import com.google.common.base.Preconditions;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * An implementation for {@link FuseFileStream} for read only operations against an Alluxio uri.
+ */
+@ThreadSafe
+public class FuseFileInStream implements FuseFileStream {
+  private final FileInStream mInStream;
+  private final long mFileLength;
+  private final AlluxioURI mURI;
+
+  /**
+   * Creates a {@link FuseFileInStream}.
+   *
+   * @param fileSystem the file system
+   * @param uri the alluxio uri
+   * @param flags the fuse create/open flags
+   * @param status the uri status, null if not uri does not exist
+   * @return a {@link FuseFileInStream}
+   */
+  public static FuseFileInStream create(FileSystem fileSystem, AlluxioURI uri,
+      int flags, Optional<URIStatus> status) {
+    Preconditions.checkNotNull(fileSystem);
+    Preconditions.checkNotNull(uri);
+    if (AlluxioFuseOpenUtils.containsTruncate(flags)) {
+      throw new UnsupportedOperationException(String.format(
+          "Failed to create read-only stream for path %s: flags 0x%x contains truncate",
+          uri, flags));
+    }
+    if (!status.isPresent()) {
+      throw new UnsupportedOperationException(String.format(
+          "Failed to create read-only stream for %s: file does not exist", uri));
+    }
+
+    URIStatus uriStatus = status.get();
+    if (!uriStatus.isCompleted()) {
+      // Cannot open incomplete file for read
+      // wait for file to complete in read or read_write mode
+      if (!AlluxioFuseUtils.waitForFileCompleted(fileSystem, uri)) {
+        throw new UnsupportedOperationException(String.format(
+            "Failed to create read-only stream for %s: incomplete file", uri));
+      }
+    }
+
+    try {
+      FileInStream is = fileSystem.openFile(uri);
+      return new FuseFileInStream(is, uriStatus.getLength(), uri);
+    } catch (IOException | AlluxioException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private FuseFileInStream(FileInStream inStream, long fileLength, AlluxioURI uri) {
+    Preconditions.checkNotNull(inStream);
+    Preconditions.checkNotNull(uri);
+    mInStream = inStream;
+    mFileLength = fileLength;
+    mURI = uri;
+  }
+
+  @Override
+  public synchronized int read(ByteBuffer buf, long size, long offset) {
+    final int sz = (int) size;
+    int nread = 0;
+    int rd = 0;
+    try {
+      if (offset - mInStream.getPos() >= mInStream.remaining()) {
```

Comment:
This is basically saying offset is larger than the file size? Can we do `offset >= mFileLength` instead?
Also we don't error out if the offset is larger than the file size?

Reply:
that's also what i am confusing, this PR mainly doing the code refactoring without changing the main read/write logic. Will create another PR for modifying the core read/write code